rclpy_examples/talker_asyncio.py

```diff
@@ -0,0 +1,33 @@
+import asyncio
+import sys
+
+import rclpy
+from rclpy.qos import qos_profile_default
+from std_msgs.msg import String
+
+
+@asincio.coroutine
+def main(args):
+    rclpy.init(args)
+
+    node = rclpy.create_node("talker")
+
+    chatter_pub = node.create_publisher(String, "chatter", qos_profile_default)
+
+    loop_rate = rclpy.WallRate(2)
+
+    msg = String()
+
+    i = 1
+    while rclpy.ok():
+      msg.data = "Hello World: " + i
+      i += 1
+      print("Publishing: '%s'" % msg.data)
+      chatter_pub.publish(msg)
+      yield from rclpy.spin_some(node)
+      yield from loop_rate.sleep()
```

Comment:
Seems like we could do something a little more `asyncio`y here. Maybe like this:

```python
import asyncio
import sys

import rclpy
from std_msgs.msg import String


@asyncio.coroutine
def publish_loop(publisher, loop_rate):
    msg = String()
    i = 1
    while rclpy.ok():
        msg.data = "Hello World: {0}".format(i)
        i += 1
        print("Publishing: '{0}'".format(msg.data))
        publisher.publish(msg)
        yield from loop_rate.sleep()


@asyncio.coroutine
def main(args):
    rclpy.init(args)
    node = rclpy.create_node("talker")
    chatter_pub = node.create_publisher(String, "chatter", rclpy.qos_profile_default)
    loop_rate = rclpy.WallRate(2)
    asyncio.ensure_future(publish_loop(chatter_pub, loop_rate))
    yield from rclpy.spin(node)

if __name__ == '__main__':
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main(sys.argv))
    loop.close()

```

The benefit here is that if `yield from rclpy.spin_some(node)` internally has `yield from`'s the `loop_rate.sleep()` can return between them. Basically we can avoid having to interleave the activities ourself and instead let asyncio do it, and more efficiently at that. You could also consider using `asyncio.wait` (https://docs.python.org/3/library/asyncio-task.html#asyncio.wait) in combination with `rclpy.spin(node)`, `loop_rate.sleep`, and `asyncio.FIRST_COMPLETED`.